common/include/opae/cxx/core/handle.h

```diff
@@ -60,33 +60,44 @@ class handle {
    */
   operator fpga_handle() const { return handle_; }
 
-  /** Write 32-bit value to MMIO.
-   * @param[in] offset The byte offset from MMIO base to write.
-   * @param[in] value  The value to be written.
-   * @return Whether the write was successful.
-   */
-  virtual bool write(uint64_t offset, uint32_t value);
-
-  /** Write 64-bit value to MMIO.
-   * @param[in] offset The byte offset from MMIO base to write.
-   * @param[in] value  The value to be written.
-   * @return Whether the write was successful.
-   */
-  virtual bool write(uint64_t offset, uint64_t value);
-
-  /** Read 32-bit value from MMIO.
-   * @param[in]  offset The byte offset from MMIO base to read.
-   * @param[out] value  Receives the value read.
-   * @return Whether the read was successful.
+  /**
+   * @brief Read CSR from resource associated with the handle
+   *
+   * @tparam T The type of value to return.
+   *
+   * @note Only uint32_t or uint64_t are allowed.
+   *
+   * @param offset The register offset
+   * @param csr_space The CSR space to read from. Default is 0.
+   *
+   * @return The value of type T read from the CSR
    */
-  virtual bool read(uint64_t offset, uint32_t &value) const;
-
-  /** Read 64-bit value from MMIO.
-   * @param[in]  offset The byte offset from MMIO base to read.
-   * @param[out] value  Receives the value read.
-   * @return Whether the read was successful.
+  template<typename T>
```

Comment:
I follow that we want the generic case to be prevented, but why don't we use = delete?
Also, don't we need the signatures of the allowed fn's (uint32_t, uint64_t) to be declared here so that clients can see them?  I might be missing something here..

Reply:
As a templated fn we can't delete it otherwise it won't compile. The signatures for uint32_t/uint64_t are template specializations and can go in the cpp. The only reason I put them in the cpp is because the mmio.h header is included in .cpp file and not the .h. I could've just as easily included mmio.h in this header and implemented the specialization there too.
You did make me think of something though. The way it's currently implemented, throwing an exception is a runtime error and not a compile time error. I think we would actually prefer to have a compile time error here which would require something like throwing in a static_assert on the template argument. Given this, and the points raised by @michael-adler, I'm starting to wonder if it's more trouble than what it's worth...

Reply:
I thought you would be able to explicitly instantiate the two forms you needed and then = delete the rest.  No matter if we don't want templates after all.

Reply:
To me it seems like more trouble than it's worth. Having the template will annoy the hardware people (who have to type more C++ syntax to specify what they want) and give the software people a false sense of security that the system will infer correctly what it should be doing when generating MMIO traffic. Using explicit names would keep the hardware people happy and keep the software folks from getting into trouble they don't understand.

Reply:
Yeah, I'll probably change it to read_csr(32|64) and write_csr(32|64).  As I said before, we're not really accomplishing any sort of generic programming. Thanks for your input 😄